[feat] add the UniT model and supporting utils

ronghanghu opened the pull request:
This PR is built upon https://github.com/facebookresearch/mmf/pull/740

- add the UniT model
- add `AdamWSkipParamsWithZeroGrad` optimizer (skips unused params)
- provide dataset zoo download for VQAv2 splits w.r.t. COCO 2017.
- provide model zoo download for UniT.

Test Plan:

Tested locally and verified the outputs

---

Evaluating the UniT model zoo entry
```
# on a 8-GPU machine
python mmf_cli/run.py \
    config=projects/unit/configs/all_8_datasets/shared_dec.yaml \
    datasets=detection_coco,detection_visual_genome,vqa2,visual_entailment,glue_qnli,glue_sst2,glue_mnli_mismatched,glue_qqp \
    model=unit \
    run_type=val \
    training.batch_size=8 \
    checkpoint.resume_zoo=unit.defaults
```

Evaluation outputs on the 8-dataset model:
```
val/detection_coco/detection_mean_ap: 0.3899,
val/detection_visual_genome/detection_mean_ap: 0.0329,
val/vqa2/vqa_accuracy: 0.6696,
val/visual_entailment/accuracy: 0.7316,
val/glue_qnli/accuracy: 0.8790,
val/glue_sst2/accuracy: 0.8922,
val/glue_mnli_mismatched/accuracy: 0.8090,
val/glue_qqp/accuracy: 0.9065
```

vedanuj commented on mmf/modules/optimizers.py:
```diff
@@ -1,7 +1,83 @@
+
+
 registry.register_optimizer("adam_w")(AdamW)
+registry.register_optimizer("adam_w_skip_params_with_zero_grad")(
```
You can just do `@registry.register_optimizer("adam_w_skip_params_with_zero_grad")` in L10

ronghanghu replied:
(fixed)

vedanuj commented on projects/unit/configs/all/shared_dec.yaml:
```diff
@@ -0,0 +1,196 @@
+      - vqa2/defaults/annotations/imdb_val2017.npy
+      test:
+      - vqa2/defaults/annotations/imdb_test2015.npy
+  # hateful_memes:
+  #   use_features: false
+  #   use_images: true
+  #   zoo_requirements: []
+  #   annotations:
+  #     train:
+  #     - /private/home/dkiela/data/datasets/hm-final-unseen/train.jsonl
+  #     val:
+  #     - /private/home/dkiela/data/datasets/hm-final-unseen/dev.jsonl
+  #     test:
+  #     - /private/home/dkiela/data/datasets/hm-final-unseen/test.jsonl
+  #   images:
+  #     train:
+  #     - /private/home/dkiela/data/datasets/hm-final-unseen
+  #     val:
+  #     - /private/home/dkiela/data/datasets/hm-final-unseen
+  #     test:
+  #     - /private/home/dkiela/data/datasets/hm-final-unseen
+  #   processors:
+  #     detr_image_processor:
+  #       type: detr_image
+  #       params:
+  #         image_size: 224
+  #         max_size: 224
```
Remove if not required

vedanuj commented on projects/unit/configs/all/shared_dec.yaml:
```diff
@@ -0,0 +1,196 @@
+    zoo_requirements: []
+    annotations:
+      train:
+      - /private/home/asg/data/visual_entailment/defaults/annotations/snli_ve_train.jsonl
+      val:
+      - /private/home/asg/data/visual_entailment/defaults/annotations/snli_ve_dev.jsonl
+      test:
+      - /private/home/asg/data/visual_entailment/defaults/annotations/snli_ve_test.jsonl
+    images:
+      train:
+      - /private/home/asg/data/visual_entailment/defaults/images/flickr30k_images
+      val:
+      - /private/home/asg/data/visual_entailment/defaults/images/flickr30k_images
+      test:
+      - /private/home/asg/data/visual_entailment/defaults/images/flickr30k_images
```
These need to change to relative paths for cache

ronghanghu replied:
(changed to relative paths)

vedanuj commented on mmf/modules/detr/models/__init__.py:
```diff
@@ -0,0 +1 @@
+# Copyright (c) Facebook, Inc. and its affiliates.
```
I think these modules are specific for Unit model only and so these can be placed under `models/unit/..`. @apsdehal what do you suggest?

ronghanghu replied:
Thanks! I'll move them to a folder under `mmf.models.unit_modules`.

ronghanghu replied:
Thanks for the suggestion! I've moved them to `mmf.models.unit`

vedanuj commented on mmf/datasets/processors/processors.py:
```diff
@@ -1493,3 +1493,29 @@ def __call__(self, item: Union[Dict[str, Any], str]) -> Dict[str, Any]:
+
+        import mmf.modules.detr.datasets.transforms as T
+
+        self.detr_im_transform = T.Compose(
+            [
+                T.RandomResize([config.image_size], max_size=self.max_size),
+                T.ToTensor(),
+                T.Normalize([0.485, 0.456, 0.406], [0.229, 0.224, 0.225]),
+            ]
+        )
```
We should also be able to remove the `detr_image` specific modifications to the datasets in the above files.

ronghanghu replied:
(this is moved to the dataset PR: https://github.com/facebookresearch/mmf/pull/740)

Since `RandomResize` is not available in torchvision_transforms, I kept the DETRImageProcessor there.

vedanuj reviewed (changes requested):
This looks awesome! Left some comments.

ronghanghu commented on mmf/common/batch_collator.py:
```diff
@@ -21,11 +21,30 @@ def __call__(self, batch):
         return sample_list
+
+
+def _build_sample_list_with_detr_fields(batch):
```
Thanks for the comments. I'll handle it.

ronghanghu replied:
(this is now resolved; the batch collator stays the same)

Also, after a discussion with Amanpreet, we decided not to support dynamic image sizes in batching, since it is not used in our UniT model or any vision SSL model, while adding a large maintenance overhead.

ronghanghu commented:
@vedanuj Thanks for the review! There is a current error in running existing checkpoints and I'm trying to figure out why. I'm working on debugging it and will incorporate these comments.

ronghanghu commented on mmf/configs/models/unit/defaults.yaml:
```diff
@@ -0,0 +1,105 @@
+      use_task_embedding_in_decoder: false
+      use_task_embedding_in_lang_encoder: false
+
+    datasets:
```
(changed to `heads` now)

ronghanghu commented:
Thanks @vedanuj for the review! I've updated this PR accordingly.

Regarding the `visual_entailment` dataset, shall we provide a download for it in MMF dataset zoo? (And if yes, should we add it in this PR?)

apsdehal commented on mmf/configs/models/unit/defaults.yaml:
```diff
@@ -0,0 +1,104 @@
+    heads:
+      detection:
+        detection_coco:
+          task_idx: 0
```
We can probably do a better job instead of incrementing. Can we auto increment?

ronghanghu replied:
On second thought, I think it is better to just manually define `task_idx` rather than automatically incrementing it, such that each dataset will have a fixed (i.e. manually specified) task_idx. This would make it easier to e.g. fine-tune a multi-task model on a single task later on, or initializing from a single task model to do multi-task training. (If it is automatically incremented, there could be mismatches in task_idx in these cases.)

apsdehal commented on mmf/models/unit/backbone.py:
```diff
@@ -0,0 +1,152 @@
+        if num_batches_tracked_key in state_dict:
+            del state_dict[num_batches_tracked_key]
+
+        super(FrozenBatchNorm2d, self)._load_from_state_dict(
```
```suggestion
        super()._load_from_state_dict(
```

apsdehal replied:
torchvision already has a FrozenBatchNorm2d. Would that work instead of reimplementing?

ronghanghu replied:
(updated to using `FrozenBatchNorm2d` from torchvision)

apsdehal commented on mmf/models/unit/backbone.py:
```diff
@@ -0,0 +1,152 @@
+        return out
+
+
+class Backbone(BackboneBase):
```
This is also in torchvision. See if can be reused.

ronghanghu replied:
(I didn't find a corresponding implementation of class `Backbone` in torchvision)

apsdehal commented on mmf/models/unit/backbone.py:
```diff
@@ -0,0 +1,152 @@
+        super().__init__(backbone, train_backbone, num_channels, return_interm_layers)
+
+
+class Joiner(nn.Sequential):
```
Same.

ronghanghu replied:
I didn't find a corresponding implementation of class `Joiner ` in torchvision. I think this is unique to DETR and not in torchvision (yet).

apsdehal commented on mmf/models/unit/backbone.py:
```diff
@@ -0,0 +1,152 @@
+        return out, pos
+
+
+def build_backbone(args):
```
same

ronghanghu replied:
(I didn't find a corresponding implementation in torchvision)

apsdehal replied:
I see, there is a Resnet specific backbone implementation.

apsdehal commented on mmf/configs/models/unit/defaults.yaml:
```diff
@@ -0,0 +1,104 @@
+        glue_mnli_mismatched: false
+        glue_qqp: false
+
+    detr_args:
```
Can we name these something else?

ronghanghu replied:
renamed to `base_args`

apsdehal commented on mmf/configs/models/unit/defaults.yaml:
```diff
@@ -0,0 +1,104 @@
+model_config:
+  unit:
+    detr_ckpt_path: ''  # the initial DETR parameters to start from (this should be a checkpoint file trained w/ DETR package)
```
Can we name this something else as well. `unit_ckpt_path` or `ckpt_path`.

ronghanghu replied:
renamed to `base_ckpt_path`

apsdehal commented on mmf/configs/models/unit/defaults.yaml:
```diff
@@ -0,0 +1,104 @@
+
+    keep_only_bert_cls:
+      # whether to keep only the [CLS] token's hidden states in transformer
+      vl:
```
`unit_ckpt_load_backbone_only` or `ckpt_load_backbone_only`.

ronghanghu replied:
renamed to `base_ckpt_load_backbone_only`

apsdehal commented on mmf/configs/models/unit/defaults.yaml:
```diff
@@ -0,0 +1,104 @@
+          task_idx: 1
+          num_classes: 1601
+          use_attr: true
+        detection_lvis:
+          task_idx: 2
+          num_classes: 1235
+          use_attr: false
```
Remove?

ronghanghu replied:
(removed)

apsdehal commented on mmf/configs/models/unit/defaults.yaml:
```diff
@@ -0,0 +1,104 @@
+        vqa2:
+          task_idx: 3
+          num_labels: 3129
+        hateful_memes:
+          task_idx: 4
+          num_labels: 2
```
Remove?

ronghanghu replied:
(removed)

apsdehal commented on mmf/models/unit/matcher.py:
```diff
@@ -0,0 +1,166 @@
+        ]
+
+
+class SequentialMatcher(nn.Module):
+    def forward(self, outputs, targets):
+        return [(torch.arange(len(tgt["labels"])),) * 2 for tgt in targets]
+
+
+class LexicographicalMatcher(nn.Module):
+    def __init__(self, lexic="acb"):
+        super().__init__()
+        self.lexic = lexic
+
+    def forward(self, outputs, targets):
+        indices = []
+        for tgt in targets:
+            tgt_cls, tgt_box = tgt["labels"], tgt["boxes"]
+            area = tgt_box[:, 2] * tgt_box[:, 3]
+            if self.lexic == "acb":
+                search_list = [
+                    (-a, cl, b)
+                    for cl, a, b in zip(
+                        tgt_cls.tolist(), area.tolist(), tgt_box.tolist()
+                    )
+                ]
+            else:
+                search_list = [
+                    (cl, -a, b)
+                    for cl, a, b in zip(
+                        tgt_cls.tolist(), area.tolist(), tgt_box.tolist()
+                    )
+                ]
+            # argsort from https://stackoverflow.com/questions/3382352/equivalent-of-numpy-argsort-in-basic-python
+            j = sorted(range(len(search_list)), key=search_list.__getitem__)
+            j = torch.as_tensor(j, dtype=torch.int64)
+            i = torch.arange(len(j), dtype=j.dtype)
+            indices.append((i, j))
+        return indices
+
+
```
Do we need others if we are only going to use HungarianMatcher always?

ronghanghu replied:
(updated to keep only HungarianMatcher)

apsdehal commented on mmf/models/unit/unit.py:
```diff
@@ -0,0 +1,335 @@
+
+        ckpt_path = self.config.detr_ckpt_path
+        if ckpt_path != "":
+            logger.info("initializing base model (DETR) from {}".format(ckpt_path))
```
```suggestion
            logger.info("initializing base model (UniT) from {}".format(ckpt_path))
```

apsdehal commented on mmf/models/unit/unit.py:
```diff
@@ -0,0 +1,335 @@
+        self.loss_calculation_fn["vl"] = self.classifier_loss_calculation
+        self.loss_calculation_fn["glue"] = self.classifier_loss_calculation
+
+        self.losses_dict = {
+            "vl": {
+                "vqa2": nn.BCEWithLogitsLoss(reduction="mean"),
+                "hateful_memes": nn.functional.cross_entropy,
+                "visual_entailment": nn.functional.cross_entropy,
+            },
+            "glue": {
+                "glue_qnli": nn.functional.cross_entropy,
+                "glue_qqp": nn.functional.cross_entropy,
+                "glue_mnli_mismatched": nn.functional.cross_entropy,
+                "glue_sst2": nn.functional.cross_entropy,
+            },
+        }
```
Can we do something to remove this hardcode?

apsdehal replied:
I probably don't want anything related to datasets hardcoded in the code.

apsdehal reviewed (commented):
Looked through half diff. Will look through other other half soon. This and previous PRs are missing typings and tests in general.

ronghanghu commented:
@apsdehal Thanks for reviewing! This PR has been updated accordingly (details in the comments above).

vedanuj commented on mmf/models/unit/unit_base_model.py:
```diff
@@ -0,0 +1,426 @@
+from .transformer import build_transformer
+
+
+ATTRIBUTE_CLASS_NUM = 401
+MAX_ATTR_NUM = 16
```
Suggestion : Should we make these configurable?

ronghanghu replied:
(added to config)

vedanuj commented on mmf/models/unit/unit.py:
```diff
@@ -0,0 +1,339 @@
+# Copyright (c) Facebook, Inc. and its affiliates.
+# Initial version was taken from https://github.com/uclanlp/visualbert
+# which was cleaned up and adapted for MMF.
```
Is this adapted from visualbert?

ronghanghu replied:
(removed this line)

vedanuj commented on projects/unit/configs/all/shared_dec.yaml:
```diff
@@ -0,0 +1,193 @@
+    base_ckpt_load_backbone_only: true
+
+dataset_config:
+  vqa2:
```
These are repeated in multiple configs. You should try to see if these can be moved to a default config which is included in other variants.

ronghanghu replied:
(moved to a default config)

vedanuj commented on projects/unit/configs/all/shared_dec.yaml:
```diff
@@ -0,0 +1,193 @@
+                std: [0.229, 0.224, 0.225]
+    images:
+      train:
+      - coco/defaults/images/trainval2014
+      - coco/defaults/images/trainval2014
+      - coco/defaults/images/trainval2014
```
COCO 2014 usually doesn't have a single `trainval2014` directory. `train2014` and `val2014` are separate directories. Please make sure this is what is extracted from the zoo downloads or else users will find it difficult to train this model.

ronghanghu replied:
(split the imdbs w.r.t. train2014 and val2014)

vedanuj reviewed (commented):
Looking good. Most of the methods are missing typings. Please add it if possible. Added a few other comments.

vedanuj commented on mmf/models/unit/position_encoding.py:
```diff
@@ -0,0 +1,106 @@
+        return pos
+
+
+def build_position_encoding(args):
```
Same for other build.. methods in this PR.

ronghanghu replied:
(Since it is building only one class of positional encoding, I directly added that class into backbone.)

ronghanghu commented on mmf/models/unit/misc.py:
```diff
@@ -0,0 +1,42 @@
+        return self.tensors, self.mask
+
+    @classmethod
+    def from_tensor_list(cls, tensor_list):
```
(typings added)

ronghanghu commented on mmf/models/unit/misc.py:
```diff
@@ -0,0 +1,42 @@
+import torch
+
+
+class NestedTensor(object):
```
(docstring added)

ronghanghu commented on mmf/models/unit/position_encoding.py:
```diff
@@ -0,0 +1,106 @@
+            scale = 2 * math.pi
+        self.scale = scale
+
+    def forward(self, tensor_list):
```
(typings added)

ronghanghu commented on mmf/models/unit/position_encoding.py:
```diff
@@ -0,0 +1,106 @@
+        return pos
+
+
+class PositionEmbeddingLearned(nn.Module):
```
(typings added)

ronghanghu commented on mmf/models/unit/matcher.py:
```diff
@@ -0,0 +1,123 @@
+        ), "all costs cant be 0"
+
+    @torch.no_grad()
+    def forward(self, outputs, targets):
```
(typings added)

ronghanghu commented:
@apsdehal @vedanuj @ytsheng This and previous PRs have been updated accordingly. Below is a summary of new changes since the last review.

- Multi-dataset support (https://github.com/facebookresearch/mmf/pull/739)
  - mostly minor changes, e.g. renaming "master_only" => "execute_on_master_only"

- Detection datasets (https://github.com/facebookresearch/mmf/pull/740)
  - move detection prediction gathering into COCO dataset's format_for_prediction
  - add type hints to functions
  - other minor changes, e.g. change to max_size = MAX_SIZE_LIMIT

- UniT model (https://github.com/facebookresearch/mmf/pull/741)
  - refactor the config to take the common ones in an include file
  - make trainval2014 image folder splitted - split the imdb of imdb_train2017.npy into those in train2014 and those in val2014; also split vg_karpathy_train
  - use flickr30k images and SNLI-VE from model zoo
  - add UniT 8-dataset models to MMF model zoo
  - add type hints to functions
  - other minor changes, e.g. docstrings to a few functions
  - verified the training locally
  ----
  
  Following-up: like @apsdehal suggested in https://github.com/facebookresearch/mmf/pull/740, a few tests should be added, such as transforms.

ronghanghu commented:
@apsdehal @vedanuj @ytsheng This and previous PRs have been updated accordingly. Below is a summary of new changes since the last review.

- Multi-dataset support (https://github.com/facebookresearch/mmf/pull/739)
  - mostly minor changes, e.g. renaming "master_only" => "execute_on_master_only"

- Detection datasets (https://github.com/facebookresearch/mmf/pull/740)
  - move detection prediction gathering into COCO dataset's format_for_prediction
  - add type hints to functions
  - other minor changes, e.g. change to max_size = MAX_SIZE_LIMIT

- UniT model (https://github.com/facebookresearch/mmf/pull/741)
  - refactor the config to take the common ones in an include file
  - make trainval2014 image folder splitted - split the imdb of imdb_train2017.npy into those in train2014 and those in val2014; also split vg_karpathy_train
  - use flickr30k images and SNLI-VE from model zoo
  - add UniT 8-dataset models to MMF model zoo
  - add type hints to functions
  - other minor changes, e.g. docstrings to a few functions
  - verified the training locally

like @apsdehal suggested in https://github.com/facebookresearch/mmf/pull/740, a few tests should be added

apsdehal commented on projects/unit/configs/all_8_datasets/shared_dec.yaml:
```diff
@@ -0,0 +1,105 @@
+      dilation: true
+      use_task_embedding_in_img_encoder: true
+      use_task_embedding_in_lang_encoder: true
+    bert_config:
```
Can we add `bert_config` to the base defaults config as well?

ronghanghu replied:
(moved `bert_config` to `base_args` and changed `detr_model` to `unit_base_model` with format_state_key)

apsdehal commented on mmf/models/unit/unit_base_model.py:
```diff
@@ -0,0 +1,460 @@
+from mmf.utils.distributed import get_world_size, is_dist_initialized
+from torch import Tensor, nn
+
+from .backbone import build_unit_convnet_backbone
```
Convert to absolute imports.

ronghanghu replied:
(converted to absolute imports)

apsdehal commented on mmf/models/unit/unit_base_model.py:
```diff
@@ -0,0 +1,460 @@
+        task_type: str = "detection",
+        dataset_name: str = "detection_coco",
+        task_idx: Optional[int] = None,
+        **kwargs,
```
Remove `**kwargs` if not needed.

ronghanghu replied:
(removed)

apsdehal reviewed (approved):
Some minor comments. Should be good to land.

ronghanghu commented on mmf/configs/models/unit/defaults.yaml:
```diff
@@ -0,0 +1,105 @@
+        glue_qqp: true
+
+    loss_on_all_hs:
+      vl:
```
(moved to a single option)

ronghanghu commented on mmf/models/unit/backbone.py:
```diff
@@ -0,0 +1,147 @@
+from torchvision.models._utils import IntermediateLayerGetter
+from torchvision.ops.misc import FrozenBatchNorm2d
+
+from .misc import NestedTensor
```
(converted to absolute imports)
